### DISCUSSION TITLE
The body_add_flextable functions slows down each time it is called

### DISCUSSION THREAD

**Discussion by moreron** [davidgohel/officer#507] [Q&A] [answered]
Been using the excellent officer and flextable libraries for a while now, so helpful. Have recently seen what I think is a new behaviour when inserting a large(ish) number of flextables into the doc object. Basically the time required to run body_add_flextable function increases each time you call it. 

This behaviour seems to be true whether calling the function in a for loop, lapply function, or manually writing each line. My current workaround is to print then read_docx after every few body_add_flextable calls otherwise the run time is in the hours not minutes. This only appears to be an issue with flextables body_add_gg and body_add_img are fine. Not sure if this is an officer or flextable package issue?

Plot from the code below to iteratively run body_add_flextable 20 times in the same doc object.
![body_add_flextable](https://github.com/davidgohel/officer/assets/8306454/edde743b-ce23-4c1d-8ad7-4bf58e5bd29e)

The code below gives representative version of the flextable, though the actual document has more complex and heavily formated ones (so can't use body_add_table). Is anyone else seeing this behaviour? Thanks for any pointers/assistence!

Session info
R version 4.3.0 (2023-04-21 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] flextable_0.9.1 officer_0.6.2  

rm(list = ls())

require(officer)
require(flextable)

#### Create sample data.frame ####
yr <- as.integer(strftime(Sys.Date(), "%Y"))
yrs <- 11
mns <- 12
cnts <- sample(0:20, size = yrs * mns, replace = T)

df <- data.frame(matrix(cnts, 
                        nrow = yrs, 
                        ncol = mns, 
                        byrow = T,
                        dimnames = list(seq(yr - 10, yr),
                                        head(month.abb, 12))))

df <- cbind(Year = rownames(df), data.frame(df, row.names = NULL))
df$Total <- rowSums(df[, 2:13], )
rm(cnts, mns, yr, yrs)

#### Create flextable ####
std_b <- fp_border(color="grey40")
ft <- flextable(df)
ft <- theme_zebra(ft, 
                   odd_header = "white", 
                   odd_body = "#7bccc4", 
                   even_body = "white" )

ft <- border_inner_h(ft, border = std_b, part = "header")
ft <- hline(ft, i = 1, border = std_b, part = "header")
ft <- vline(ft, j = c("Year", "Dec"), border = std_b, part = "all")
ft <- width(ft, width = c(1.35, rep(1.2, 12), 1.35), unit = "cm")
rm(std_b)

#### Create new doc with titles, bookmarks and page breaks ####
titles <- paste("Flextable", 1:15)

if(exists("doc")) rm(doc)

doc <- read_docx()

for(t in titles) {
  
  doc <- body_add_par(doc, t, style = "heading 3")
  doc <- body_bookmark(doc, tolower(sub("Flextable ", "ft", t)))
  doc <- body_add_break(doc)  
  
}

rm(t)

#### Add flextables to doc bookmarks using for loop ####
tt_ft <- data.frame()
starter <- Sys.time()

for(t in titles) {

  b <- Sys.time()
  doc <- body_add_flextable(cursor_bookmark(doc, tolower(sub("Flextable ", 
                                                             "ft", t))),
                            value = ft,
                            align = "left", 
                            pos = "after")   
  s <- Sys.time() - b
  tt <- data.frame(obj = t, secs = as.numeric(s))
  tt_ft <- rbind(tt_ft, tt)
  print(s)
}
Sys.time() - starter

#### Plot time taken for each iteration ###
plot(tt_ft$secs, xlab = "iteration: body_add_flextable", ylab = "seconds taken")

#### print(doc, target = "example_flextable.docx")  ####

rm(list = ls())

**Comment by davidgohel** [comment, marked as answer]
we are using package xml2 and xml manipulation operate on RAM, tables in Word are big XML structures. You can use R Markdown as a substitute if you need huge reports, the behavior will be slower for small doc but faster for huge doc with tables
